fix: update cert secret if exists

busser opened the pull request:
Fixes #378.

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/vault-k8s?pullRequest=401)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/vault-k8s?pullRequest=401)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/vault-k8s?pullRequest=401).</sub>

busser commented:
@tvoran no rush, but if you have time to take a look that would be great 🙂

tvoran commented on helper/cert/source_gen.go:
```diff
@@ -236,25 +236,31 @@ func (s *GenSource) retryUpdateSecret(ctx context.Context, leaderCh chan bool, b
-	// Attempt updating the Secret first, and if it doesn't exist, fallback to
-	// create
-	_, err := s.K8sClient.CoreV1().Secrets(s.Namespace).Update(ctx, secret, metav1.UpdateOptions{})
+	secret, err := s.K8sClient.CoreV1().Secrets(s.Namespace).Get(ctx, certSecretName, metav1.GetOptions{})
```
I think we may want to do the get from the informer cache here, instead of the regular k8s client. That should help avoid any rate limiting from the k8s API, even though this get shouldn't be firing that often. Something like:

```suggestion
	secret, err := s.SecretsCache.Lister().Secrets(s.Namespace).Get(certSecretName)
```

Similar to getBundleFromSecret(): https://github.com/hashicorp/vault-k8s/blob/0603c2a03300773fd6ba6174c6d4327d70889d58/helper/cert/source_gen.go#L266-L269

busser replied:
It's done 👍

tvoran commented:
Looks like we'll also need to update the tests, since `make test` crashes in `TestGenSource_leader()`. I bet the secret informer will need to be setup in that test, similar to `TestGenSource_follower()`.

It would also be great to add another test that starts with an empty secret with labels and annotations, runs the cert generation, then checks that the updated secret's labels and annotations are intact. Maybe another variant of `TestGenSource_leader()`?

busser commented:
`TestGenSource_leader()` now fixed. Like you said, it just needed the informer to be setup.

I added a `TestGenSource_leader_expiry()`, which combines logic from `TestGenSource_leader` and `TestGenSource_expiry` to confirm that the secret retains its metadata when it is updated with a new bundle.

tvoran reviewed (approved):
Thanks for adding the tests!

kenske commented:
Great work, thanks! Any idea when this will get released?

tvoran commented:
@kenske Sometime later this month is the best estimate I can give right now.
